homeassistant/components/apple_tv/config_flow.py

```diff
@@ -0,0 +1,442 @@
+from pyatv import conf, const, convert, exceptions, pair, scan
+
+from .const import (CONF_CREDENTIALS, CONF_CREDENTIALS_AIRPLAY,
+                    CONF_CREDENTIALS_DMAP, CONF_CREDENTIALS_MRP,
+                    CONF_IDENTIFIER, CONF_START_OFF, DOMAIN)
+
+_LOGGER = logging.getLogger(__name__)
+
+INPUT_PIN_SCHEMA = vol.Schema({vol.Required(CONF_PIN, default=None): int})
+
+DEFAULT_START_OFF = False
+PROTOCOL_PRIORITY = [const.Protocol.MRP, const.Protocol.DMAP, const.Protocol.AirPlay]
+
+# Mapping between config entry format and pyatv
+CREDENTIAL_MAPPING = {
+    CONF_CREDENTIALS_MRP: const.Protocol.MRP.value,
+    CONF_CREDENTIALS_DMAP: const.Protocol.DMAP.value,
+    CONF_CREDENTIALS_AIRPLAY: const.Protocol.AirPlay.value,
+}
+
+
+async def device_scan(identifier, loop, cache=None):
+    def _filter_device(dev):
+        if identifier is None:
+            return True
+        if identifier == str(dev.address):
+            return True
+        if identifier == dev.name:
+            return True
+        return any([x.identifier == identifier for x in dev.services])
+
+    def _host_filter():
+        try:
+            return [ip_address(identifier)]
+        except ValueError:
+            return None
+
+    if cache:
+        matches = [atv for atv in cache if _filter_device(atv)]
+        if matches:
+            return cache, matches[0]
+
+    for hosts in [_host_filter(), None]:
+        scan_result = atvs = await scan(loop, timeout=3, hosts=hosts)
+        matches = [atv for atv in scan_result if _filter_device(atv)]
+
+        if matches:
+            return scan_result, matches[0]
+
+    return scan_result, None
+
+
+def is_valid_credentials(credentials):
+    return (
+        credentials.get(const.Protocol.MRP.value) is not None
+        or credentials.get(const.Protocol.DMAP.value) is not None
+    )
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class AppleTVConfigFlow(config_entries.ConfigFlow):
+    """Handle a config flow for Apple TV."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get options flow for this handler."""
+        return AppleTVOptionsFlow(config_entry)
+
+    def __init__(self):
+        """Initialize a new AppleTVConfigFlow."""
+        self.scan_result = None
+        self.atv = None
+        self.identifier = None
+        self.protocol = None
+        self.pairing = None
+        self.credentials = {}  # Protocol -> credentials
+
+    async def async_step_invalid_credentials(self, info):
+        """Handle initial step when updating invalid credentials."""
+        self.identifier = info.get(CONF_IDENTIFIER)
+        self.context["title_placeholders"] = {"name": info.get(CONF_NAME)}
+
+        for flow in self._async_in_progress():
+            if flow["context"].get("identifier") == self.identifier:
+                return self.async_abort(reason="already_configured")
```

Comment:
await self.async_set_unique_id(self.identifier)

Reply:
Didn't know about that. I should probably adjust other code paths according to this as well, but I'll leave that for a future improvement. Will update here now though.